src/cudamatrix/cu-kernels.cu

```diff
@@ -2599,6 +2599,60 @@ static void _diff_softmax(Real* x, const MatrixDim dim, const Real* value,
   }
 }
 
+// Differentiate backward through the log softmax function.
+// "value" is the log softmax output. Does, for each row i,
+// x(i) =  diff(i) - sum(diff(i)) .* exp(value(i))
+// ???(i) is row-vector.
+// CUDA thread layout: one thread block (256 threads) per matrix-row.
+template<typename Real>
+__global__
+static void _diff_log_softmax(Real* x, const MatrixDim dim, const Real* value,
```

Comment:
It is now following the naming in `nnet3` as `in_deriv`, `out_value` and `out_deriv`. The output param is put to the end, but we may need to change all other kernel functions later.